### PR TITLE
Load the default config file before working with default properties

### LINKED ISSUE #12565: EVP_default_properties_is_fips_enabled() not working as expected
This issue is reproduced from the following openssl-users post:

https://mta.openssl.org/pipermail/openssl-users/2020-July/012811.html

I'm struggling to understand how `EVP_default_properties_is_fips_enabled()` works. I cannot get this function to return nonzero unless I first call either `EVP_default_properties_enable_fips()` or `EVP_set_default_properties()`, even when the config file sets default_properties to enable fips.

Also, the return value of this function doesn't seem to have any effect on which provider gets selected (which I think is what issue #11594 describes?).

My config file has the following:

````
[openssl_init]
providers = provider_sect
alg_section = alg_sect

[provider_sect]
fips = fips_sect
default = default_sect

[default_sect]
activate = 1

[alg_sect]
default_properties = fips=yes

.include /path/to/fips.cnf
````

I understand this to mean both the default provider and the fips provider will be loaded into the default context, and both of these providers will be activated. I also see that:

````
EVP_MD_fetch(NULL, "sha256", NULL);
````

returns a pointer which `EVP_MD_provider()` confirms as being from the fips provider (as expected). Changing this to "fips=no" in the config file results in `EVP_MD_fetch()` returning `EVP_MD` from the default provider, again as expected. However, in both cases, `EVP_default_properties_is_fips_enabled()` always returns zero. I don't see anything in #11594 that would explain this.

Calling `EVP_default_properties_enable_fips(NULL, 1)` results in `EVP_default_properties_is_fips_enabled() `returning 1, but this does not appear to override the `fips=no` from the config file during EVP_MD_fetch() (which is what I believe #11594 describes).

Is the result of `EVP_default_properties_is_fips_enabled()` supposed to take into account the default properties specified in the config file? I don't see it doing that. Also, regarding #11594, if default properties are currently still broken, why do those in the config appear to work properly?

And finally the burning question: Any ETA on a fix? :-) :-) :-)

### PR DESCRIPTION
**Pull request by mattcaswell**
A config file can change the global default properties. Therefore we
must ensure that the config file is loaded before reading or amending
them.

Fixes #12565

Marked as WIP because it needs some kind of test.

### CODE REVIEW

**Comment by levitte** [comment]
Why is the load config flag needed?

**Comment by mattcaswell** [comment]
> Why is the load config flag needed?

Because without it the config file doesn't get loaded! ;-)

**Comment by mattcaswell** [comment]
I've now updated this with a test and taken it out of WIP.

Please take a look.

File: crypto/evp/evp_fetch.c [hunk 1]
```diff
@@ -385,21 +386,26 @@ static int evp_set_default_properties(OPENSSL_CTX *libctx,
     return 0;
 }
 
-int EVP_set_default_properties(OPENSSL_CTX *libctx, const char *propq)
+int evp_set_default_properties_intern(OPENSSL_CTX *libctx, const char *propq,
+                                      int loadconfig)
```
Comment (t8m) [inline crypto/evp/evp_fetch.c:390]: It is really confusing to have evp_set_default_properties EVP_set_default_properties and evp_set_default_properties_intern. Perhaps rename evp_set_default_properties to evp_set_parsed_default_properties and (optionally) evp_set_default_properties_intern to evp_set_default_properties.

Comment (mattcaswell) [reply crypto/evp/evp_fetch.c:390]: I renamed `evp_set_default_properties` to `evp_set_parsed_default_properties` but left `evp_set_default_properties_intern` as it is.

Comment (t8m) [reply crypto/evp/evp_fetch.c:390]: One reason (unwritten here, my bad) why I suggested the rename of the _intern function was that the suffix is not used anywhere else - we usually use _int. But anyway it is fine to keep it as is.

Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: I renamed it to use _int

**Comment by mattcaswell** [comment]
I've rebased this and completely reworked the test so that it is a new standalone test rather than a part of evp_test. Previously it relied on the fact that evp_test used the default libctx. Meanwhile @slontis came along and changed it so that it didn't any more :-)

Please take another look.

File: test/recipes/30-test_defltfips.t [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                '-out', bldtop_file('providers', 'fipsmodule.cnf'),
+                '-module', $infile])),
+       "fipsinstall");
+
+    $ENV{OPENSSL_CONF} = abs_path(srctop_file("test", "fips.cnf"));
+    ok(run(test(["defltfips_test", "fips"])), "running defltfips_test fips");
+}
+
+$ENV{OPENSSL_CONF} = abs_path(srctop_file("test", "default.cnf"));
+ok(run(test(["defltfips_test"])), "running defltfips_test");
```
Comment (t8m) [inline test/recipes/30-test_defltfips.t:43]: Missing EOL at EOF.

File: test/defltfips_test.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        /* fall through */
+    default:
+        TEST_error("Invalid argument\n");
+        return 0;
+    }
+
+    /* Must be the first test before any other libcrypto calls are made */
+    ADD_TEST(test_is_fips_enabled);
+    return 1;
+}
```
Comment (t8m) [inline test/defltfips_test.c:81]: Missing EOL at EOF.

**Comment by mattcaswell** [comment]
Two missing EOLs fixed. Please take another look.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**